File: src/main/java/edu/cornell/cs/apl/viaduct/Host.java
package edu.cornell.cs.apl.viaduct;

import edu.cornell.cs.apl.viaduct.security.Label;

/** describes an abstract machine / principal with a given trust label. */
public class Host {
  private static final Host defaultHost = new Host("__DEFAULT__", Label.bottom());
  final String name;
  final Label label;

  public Host(String name) {
    this.name = name;
    this.label = Label.bottom();
  }

  public Host(String name, Label label) {
    this.name = name;
    this.label = label;
  }

  public static Host getDefault() {
    return defaultHost;
  }

  public String getName() {
    return this.name;
  }

  public Label getLabel() {
    return this.label;
  }

  @Override
  public boolean equals(Object o) {
    if (o == null) {
      return false;
    }

    if (o instanceof Host) {
      Host ohost = (Host) o;
      return this.name.equals(ohost.name) && this.label.equals(ohost.label);

    } else {
      return false;
    }
  }

  @Override
  public int hashCode() {
    return this.name.hashCode();
  }

  @Override
  public String toString() {
    return this.name;
  }
}

File: kotlin-dsl/src/main/kotlin/edu/cornell/cs/apl/viaduct/kotlin/Host.kt
package edu.cornell.cs.apl.viaduct.kotlin

import kotlin.properties.PropertyDelegateProvider
import kotlin.properties.ReadOnlyProperty

class Host internal constructor(internal val displayName: String, val label: Label)

fun host(label: Label?) = PropertyDelegateProvider { _: Any?, property ->
    val displayName = property.name
    val result = Host(displayName, label ?: Principal(displayName))
    ReadOnlyProperty<Any?, Host> { _, _ -> result }
}
